### PR TITLE
cryptenroll: change class in provided PKCS#11 URI if necessary

### PR DESCRIPTION
**Pull request by H5117**
Fixes https://github.com/systemd/systemd/issues/23479.

### CODE REVIEW

File: src/cryptenroll/cryptenroll-pkcs11.c [hunk 1]
```diff
@@ -62,12 +86,16 @@ int enroll_pkcs11(
         if (asprintf(&keyslot_as_string, "%i", keyslot) < 0)
                 return log_oom();
 
+        r = uri_set_private_class(uri, &private_uri);
```
Comment (poettering) [inline src/cryptenroll/cryptenroll-pkcs11.c:91]: can you add a comment here, that explains why this is needed

**Comment by poettering** [comment]
the last commit should have an explanation in the commit msg, and be the one that carries "Fixes: #23479"

File: src/cryptenroll/cryptenroll-pkcs11.c [hunk 2]
```diff
@@ -62,12 +86,18 @@ int enroll_pkcs11(
...
         r = json_build(&v,
-                       JSON_BUILD_OBJECT(
-                                       JSON_BUILD_PAIR("type", JSON_BUILD_CONST_STRING("systemd-pkcs11")),
-                                       JSON_BUILD_PAIR("keyslots", JSON_BUILD_ARRAY(JSON_BUILD_STRING(keyslot_as_string))),
-                                       JSON_BUILD_PAIR("pkcs11-uri", JSON_BUILD_STRING(uri)),
-                                       JSON_BUILD_PAIR("pkcs11-key", JSON_BUILD_BASE64(saved_key, saved_key_size))));
+                JSON_BUILD_OBJECT(
+                        JSON_BUILD_PAIR("type", JSON_BUILD_CONST_STRING("systemd-pkcs11")),
+                        JSON_BUILD_PAIR("keyslots", JSON_BUILD_ARRAY(JSON_BUILD_STRING(keyslot_as_string))),
+                        JSON_BUILD_PAIR("pkcs11-uri", JSON_BUILD_STRING((private_uri)? private_uri : uri)),
```
Comment (poettering) [inline src/cryptenroll/cryptenroll-pkcs11.c]: please rewrite this as:

```c
… JSON_BUILD_STRING(private_uri ?: uri) …
```

i.e. we use the GNU C extension that allows suppressing the second part of the ternary op if it is the same as the first part.

**Comment by poettering** [comment]
lgtm. just one minor tweak please

**Comment by H5117** [comment]
I added some changes in man/systemd-cryptenroll.xml.

**Pull request merged**